src/transformers/integrations/bitsandbytes.py

```diff
@@ -322,3 +325,122 @@ def get_keys_to_not_convert(model):
         filtered_module_names.append(name)
 
     return filtered_module_names
+
+
+# Copied from PEFT: https://github.com/huggingface/peft/blob/47b3712898539569c02ec5b3ed4a6c36811331a1/src/peft/utils/integrations.py#L41
+def dequantize_bnb_weight(weight: torch.nn.Parameter, state=None):
+    """
+    Helper function to dequantize 4bit or 8bit bnb weights.
+
+    If the weight is not a bnb quantized weight, it will be returned as is.
+    """
+    if not isinstance(weight, torch.nn.Parameter):
+        raise TypeError(f"Input weight should be of type nn.Parameter, got {type(weight)} instead")
+
+    cls_name = weight.__class__.__name__
+    if cls_name not in ("Params4bit", "Int8Params"):
+        return weight
+
+    import bitsandbytes as bnb
+
+    if cls_name == "Params4bit":
+        return bnb.functional.dequantize_4bit(weight.data, weight.quant_state)
```

Comment:
The user might want to know in which precision the model was dequantized since they don't have the possibility to control that. I think it could be great to give that information since there is no default value (as opposed to `from_pretrained` which loads the model in fp32). 
Two ways to get that: 
- just check the dtype of the weights at the end ( potentially the easiest way ) 
- check that happens in `dequantize_4bit` . In the method, you see that they get the output dtype with weight.quant_state.dtype. 

We can potentially add a `torch_dtype` attribute in the future if it makes sense.

Reply:
+1

Reply:
Nice catch! The output dtype should be correctly inferred here: https://github.com/TimDettmers/bitsandbytes/blob/b891f80ba514833f41f0e9226983b02a9fb5c44b/bitsandbytes/functional.py#L1349 through the compute_dtype so it should be accurate - I added a `warning_once` staement to inform users on the dequantized dtype: [1a4a906](https://github.com/huggingface/transformers/pull/30806/commits/1a4a906741c2be1bab39538d9ad117e12d3d67f9)